File: Squares.java
import java.io.*;
import java.util.*;

public class Squares
{
    public static void warn(String s) {
        System.out.println(s);
    }

    public static int _calc(int n) {
       
        int root = (int)Math.sqrt(n);

        if (n == root * root) {
            return 1;
        }

        int true_square = root * root;
        int min = Integer.MAX_VALUE;

        for (int i = 1; i*i <= true_square; i++) {
            int tmp = 1 + calc(n - i*i);

            if (tmp < min) {
                min = tmp;
            }
        }
        
        return min;
    }
    
    static int[] cache;

    public static int calc(int n) {

        if (n == 0)
            return 0;

        if (cache[n-1] == 0) {
            cache[n-1] = _calc(n);
        }

        return cache[n-1];
    }

    public static void main(String[] argv) {
        Scanner in = new Scanner(System.in);
        PrintWriter out = new PrintWriter(System.out);

        int n = in.nextInt();

        cache = new int[n];

        for (int i = 1; i < n; i++) {
            calc(i);
            //warn ("calc("+i+")="+calc(i));
        }

        out.println( "" + calc(n) );
        out.flush();
    }
}

File: squares.pl
use strict;
use warnings;

use Memoize;
memoize('calc');

sub calc
{
    my ($n) = @_;

    if ($n == int(sqrt($n))**2) {
#        warn "$n is aquared! ".sqrt($n);
        return 1;
    }
    
    my $min = 'inf';

    for (my $i = 1; $i < int($n/2); $i++) {
        my $count = calc($i) + calc($n-$i);
        if ($count < $min) {
            $min = $count;
        }
    }
    
    return $min;
}

my $n = 99;

for (1..$n) {
    calc($_);
    warn "c($_)=".calc($_);
}

warn calc($n);
